src/memory.rs

```diff
@@ -432,13 +392,13 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
     }
     
     // Returns a dtor and its argument, if one is supposed to run
-    pub(crate) fn fetch_tls_dtor(&mut self) -> Option<(ty::Instance<'tcx>, Pointer)> {
-        for (_, &mut TlsEntry { ref mut data, dtor }) in self.thread_local.iter_mut() {
-            if !data.is_null_ptr() {
-                if let Some(dtor) = dtor {
-                    let old_data = *data;
-                    *data = Pointer::from_int(0);
-                    return Some((dtor, old_data));
+    pub(crate) fn fetch_tls_dtor(&mut self) -> Option<(ty::Instance<'tcx>, PrimVal)> {
+        for (_, &mut TlsEntry { ref mut data, ref mut dtor }) in self.thread_local.iter_mut() {
+            if let Some(dtor) = dtor.take() {
```

Comment:
However, we may be implementing the ordering incorrectly... I think currently we run the first dtor multiple times (if data is set back to non-NULL), instead of running the other dtors first. Seems I missed that.